docs/design/map/map-store-operation-offloading.md

```diff
@@ -0,0 +1,247 @@
+# Offloading MapStore MapLoader Operations
+
+| ℹ️ Since: 5.2 |
+|---------------|
+
+## Summary
+Blocked partition threads limits the throughput of cluster.
+By implementing MapStore/MapLoader operation offloading,
+next operations will not be blocked in a partition.
+
+## Goals
+- Having MapStore and MapLoader operations offloaded for IMap
+- To make this offloading feature configurable per map basis.
+
+## Non-Goals
+- No offloading for `MapLoaderLifecycleSupport#init`, since it
+  is a one time operation per node along the lifecycle of a map.  
+- No offloading for tiered store operations
+- No changes in cluster wide map bulk loading mechanism.
+- JCache CacheLoader and CacheWriter are out of scope.
+
+## Description
+In its essence a [Head-of-line
+blocking](https://en.wikipedia.org/wiki/Head-of-line_blocking)
+issue we are trying to solve here. One MapStore operation can
+block other operations from completing by blocking a partition
+thread indefinitely. Since partition threads are one of the most
+important shared resources in a cluster, an execution in these
+threads must be fast so next executions can complete fast as well.
+
+Some examples of problematic cases to be addressed here:
+- After a cache miss, loading a value from MapStore  
+- Bulk loading and storing with loadAll/storeAll methods
+- When DB is down cluster is hanging
+
+### Design
+
+#### Overall View
+When MapStore interaction is needed, interaction is executed
+asynchronously and the partition thread is freed. Once
+the interaction completes, the task/operation is rescheduled on
+the original partition thread and completes. So all the other 
+logic that updates the internals remains same.
+
+#### Configuration
+##### Configuring Offload Per Map 
+`offload` field was added in `MapStoreConfig`. It's `true` by default.    
+```java
+MapStoreConfig#setOffload
+```
+```xml
+<map-store enabled="true" initial-mode="LAZY">
+    <offload>true</offload>
+    <class-name>com.hazelcast.examples.DummyStore</class-name>
+    <write-delay-seconds>60</write-delay-seconds>
+    <write-batch-size>1000</write-batch-size>
+    <write-coalescing>true</write-coalescing>
+</map-store>
+```
+```yaml
+map-store:
+  enabled: true
+  initial-mode: LAZY
+  class-name: com.hazelcast.examples.DummyStore
+  offload: true
+  write-delay-seconds: 60
+  write-batch-size: 1000
+  write-coalescing: true
+```
+##### Configuring Offloadable Executor
```

Comment:
It is same with default executor currently in use, max worker count is 16 and queue capacity is int-max. Burden is on user to find a better setting for the use case.
Regarding blocking case, yes it is possible and this is better than the current situation since partition threads are not blocked.